numpy/core/src/multiarray/iterators.c

```diff
@@ -640,34 +640,32 @@ iter_subscript(PyArrayIterObject *self, PyObject *ind)
         obj = ind;
     }
 
-    if (PyArray_Check(obj)) {
-        /* Check for Boolean object */
-        if (PyArray_TYPE((PyArrayObject *)obj) == NPY_BOOL) {
-            ret = iter_subscript_Bool(self, (PyArrayObject *)obj);
-            Py_DECREF(indtype);
-        }
-        /* Check for integer array */
-        else if (PyArray_ISINTEGER((PyArrayObject *)obj)) {
-            PyObject *new;
-            new = PyArray_FromAny(obj, indtype, 0, 0,
-                              NPY_ARRAY_FORCECAST | NPY_ARRAY_ALIGNED, NULL);
-            if (new == NULL) {
-                goto fail;
-            }
-            Py_DECREF(obj);
-            obj = new;
-            new = iter_subscript_int(self, (PyArrayObject *)obj);
-            Py_DECREF(obj);
-            return new;
-        }
-        else {
-            goto fail;
-        }
+    /* Any remaining valid input is an array or has been turned into one */
+    if (!PyArray_Check(obj)) {
+        goto fail;
+    }
+
+    /* Check for Boolean array */
+    if (PyArray_TYPE((PyArrayObject *)obj) == NPY_BOOL) {
+        ret = iter_subscript_Bool(self, (PyArrayObject *)obj);
+        Py_DECREF(indtype);
         Py_DECREF(obj);
         return (PyObject *)ret;
     }
-    else {
-        Py_DECREF(indtype);
+
+    /* Check for integer array */
+    if (PyArray_ISINTEGER((PyArrayObject *)obj)) {
+        PyObject *new;
+        new = PyArray_FromAny(obj, indtype, 0, 0,
+                          NPY_ARRAY_FORCECAST | NPY_ARRAY_ALIGNED, NULL);
+        if (new == NULL) {
+            goto fail;
```

Comment:
This causes a double decreed, the reference to indtype was stolen on line 659

Reply:
Indeed, you're right! Just to be clear, this was there [in the original too](https://github.com/numpy/numpy/blob/251c526672f897543176b966229ecde96bb89a9d/numpy/core/src/multiarray/iterators.c#L651-L655), right? I did have an uncertain feeling about that, but I was still confused about some bits of the code when I noticed that, and then I forgot.

The question is, can it happen that we have an integer array but the conversion on (the new) line 659 fails? That's the only case when this problem arises, and I should add that as a test case. But in order for this to be foolproof I'll have to increment the refcount, and decrement it if we succeeded. Thank you for spotting it!

Reply:
I think the ref should be stolen unconditionally, whether `new` succeeds or not. So your latest revision still double-decrefs `indtype`

Reply:
I'm probably missing something still. But as I see it, we [start with one reference to `indtype`](https://github.com/numpy/numpy/blob/64445bcc9ec8926f828bc109d103eb0f0d001e38/numpy/core/src/multiarray/iterators.c#L630-L633), then [bump up the refcount of `indtype`](https://github.com/numpy/numpy/blob/64445bcc9ec8926f828bc109d103eb0f0d001e38/numpy/core/src/multiarray/iterators.c#L659) because `PyArray_FromAny` steals one, so we'll still have one to spare after the call. Then either the call succeeds and we get rid of that spare reference manually before returning,
```
        Py_INCREF(indtype);
        new = PyArray_FromAny(obj, indtype, 0, 0,
                          NPY_ARRAY_FORCECAST | NPY_ARRAY_ALIGNED, NULL);
        if (new == NULL) {
            goto fail;
        }
        Py_DECREF(indtype);
        /* ... */
        return new;
```
or we jump to `fail` and let `XDECREF` do it:
```
 fail:
    /* ... */
    Py_XDECREF(indtype);
```
So in both cases we should have [from line 659](https://github.com/numpy/numpy/blob/64445bcc9ec8926f828bc109d103eb0f0d001e38/numpy/core/src/multiarray/iterators.c#L659) two references to `indtype`, one of which is stolen by `...FromAny` and one which is disposed of manually. Or, conversely, we only need to bump the refcount of `indtype` before the call to `...FromAny` because the `fail` mode includes a call to `XDECREF`.

Reply:
You're right, I missed the incref